server/datastore/mysql/software.go

```diff
@@ -71,31 +61,223 @@ func uniqueStringToSoftware(s string) fleet.Software {
 func softwareSliceToMap(softwares []fleet.Software) map[string]fleet.Software {
 	result := make(map[string]fleet.Software)
 	for _, s := range softwares {
-		result[softwareToUniqueString(s)] = s
+		result[s.ToUniqueStr()] = s
 	}
 	return result
 }
 
-// UpdateHostSoftware updates the software list of a host.
-// The update consists of deleting existing entries that are not in the given `software`
-// slice, updating existing entries and inserting new entries.
-func (ds *Datastore) UpdateHostSoftware(ctx context.Context, hostID uint, software []fleet.Software) error {
+func (ds *Datastore) UpdateHostSoftware(ctx context.Context, hostID uint, software []fleet.Software) (*fleet.UpdateHostSoftwareDBResult, error) {
+	var result *fleet.UpdateHostSoftwareDBResult
+	err := ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
+		r, err := applyChangesForNewSoftwareDB(ctx, tx, hostID, software, ds.minLastOpenedAtDiff)
+		result = r
+		return err
+	})
+	return result, err
+}
+
+func (ds *Datastore) UpdateHostSoftwareInstalledPaths(
+	ctx context.Context,
+	hostID uint,
+	reported map[string]struct{},
+	mutationResults *fleet.UpdateHostSoftwareDBResult,
+) error {
+	currS := mutationResults.CurrInstalled()
+
+	hsip, err := ds.getHostSoftwareInstalledPaths(ctx, hostID)
+	if err != nil {
+		return err
+	}
+
+	toI, toD, err := hostSoftwareInstalledPathsDelta(hostID, reported, hsip, currS)
+	if err != nil {
+		return err
+	}
+
+	if len(toI) == 0 && len(toD) == 0 {
+		// Nothing to do ...
+		return nil
+	}
+
 	return ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
-		return applyChangesForNewSoftwareDB(ctx, tx, hostID, software, ds.minLastOpenedAtDiff)
+		if err := deleteHostSoftwareInstalledPaths(ctx, tx, toD); err != nil {
+			return err
+		}
+
+		if err := insertHostSoftwareInstalledPaths(ctx, tx, toI); err != nil {
+			return err
+		}
+
+		return nil
 	})
 }
 
+func (ds *Datastore) getHostSoftwareInstalledPaths(
+	ctx context.Context,
+	hostID uint,
+) (
+	[]fleet.HostSoftwareInstalledPath,
+	error,
+) {
+	stmt := `
+		SELECT t.id, t.host_id, t.software_id, t.installed_path
+		FROM host_software_installed_paths t
+		WHERE t.host_id = ?
+	`
+
+	var result []fleet.HostSoftwareInstalledPath
+	if err := sqlx.SelectContext(ctx, ds.reader, &result, stmt, hostID); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+// hostSoftwareInstalledPathsDelta returns what should be inserted and deleted to keep the
+// 'host_software_installed_paths' table in-sync with the osquery reported query results.
+// 'reported' is a set of 'installed_path-software.UniqueStr' strings, built from the osquery
+// results.
+// 'stored' contains all 'host_software_installed_paths' rows for the given host.
+// 'hostSoftware' contains the current software installed on the host.
```

Comment:
Just making sure that I understand.
`reported` - comes from the latest query
`stored` and `hostSoftware` come from the existing database. (What's currently stored for the host)

right?
Consider better names ?

Reply:
Maybe something like:

reported   --> `listOfSWFromLatestQuery`
stored   --> `listOfSWFromDB`
hostSoftware  --> ? (Not sure I understand what this param holds.

Reply:
Sure, hope this helps:

- `reported` is the set of software installed paths reported from osquery query results.
- `stored` Are all 'host_software_installed_paths' rows for the current host
- `hostSoftware' is all the software currently installed on the host.

It wouldn't make sense to do the suggested renaming because `reported` nor `stored` hold software records.

Reply:
So `stored` is `storedPaths` ?
And `hostSoftware` is the stored list of software (Without the paths) ?